File: src/com/krypton/updater/model/repos/UpdateRepository.java
/*
 * Copyright (C) 2021 AOSP-Krypton Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.krypton.updater.model.repos;

import static android.os.UserHandle.SYSTEM;
import static com.krypton.updater.util.Constants.INDETERMINATE;
import static com.krypton.updater.util.Constants.PAUSED;
import static com.krypton.updater.util.Constants.FAILED;
import static com.krypton.updater.util.Constants.FINISHED;
import static com.krypton.updater.util.Constants.UPDATE_PENDING;
import static com.krypton.updater.util.Constants.UPDATING;

import android.content.Context;
import android.net.Uri;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.krypton.updater.model.data.DataStore;
import com.krypton.updater.model.data.OTAFileManager;
import com.krypton.updater.model.data.ProgressInfo;
import com.krypton.updater.model.data.UpdateManager;
import com.krypton.updater.model.data.UpdateStatus;
import com.krypton.updater.R;

import io.reactivex.rxjava3.processors.BehaviorProcessor;

import java.io.InputStream;
import java.io.IOException;
import java.util.concurrent.ExecutorService;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class UpdateRepository {
    private static final String TAG = "UpdateRepository";
    private final Context context;
    private final ExecutorService executor;
    private final UpdateManager updateManager;
    private final OTAFileManager ofm;
    private final DataStore dataStore;

    @Inject
    public UpdateRepository(Context context, UpdateManager updateManager,
            ExecutorService executor, OTAFileManager ofm,
            DataStore dataStore) {
        this.context = context;
        this.executor = executor;
        this.updateManager = updateManager;
        this.ofm = ofm;
        this.dataStore = dataStore;
    }

    public void startUpdate() {
        executor.execute(() -> updateManager.start());
    }

    public void pauseUpdate(boolean pause) {
        executor.execute(() -> {
            if (updateManager.isUpdating()) {
                updateManager.pause(pause);
            }
        });
    }

    public void cancelUpdate() {
        executor.execute(() -> {
            if (updateManager.isUpdating()) {
                updateManager.cancel();
                dataStore.setGlobalStatus(UPDATE_PENDING);
            }
        });
    }

    public BehaviorProcessor<UpdateStatus> getUpdateStatusProcessor() {
        return updateManager.getUpdateStatusProcessor();
    }

    public void setupLocalUpgrade(String fileName, Uri uri) {
        executor.execute(() -> {
            try (InputStream inStream = context.getContentResolver().openInputStream(uri)) {
                if (ofm.copyToOTAPackageDir(inStream)) {
                    dataStore.setLocalUpgradeFileName(fileName);
                    dataStore.setGlobalStatus(UPDATE_PENDING);
                }
            } catch (IOException e) {
                Log.e(TAG, "IOException when copying file from uri " + uri.toString(), e);
            }
        });
    }

    public ProgressInfo getProgressInfo(UpdateStatus updateStatus) {
        String status = "";
        switch (updateStatus.getStatusCode()) {
            case INDETERMINATE:
                status = getString(R.string.waiting);
                break;
            case UPDATING:
                if (updateStatus.getStep() == 1) {
                    status = getString(R.string.processing_payload);
                } else if (updateStatus.getStep() == 2) {
                    status = getString(R.string.applying_update);
                }
                break;
            case PAUSED:
                status = getString(R.string.update_paused);
                break;
            case FINISHED:
                status = getString(R.string.update_finished);
                break;
            case FAILED:
                status = getString(R.string.update_failed);
        }
        status = String.format("%s\t\t%d%%", status, updateStatus.getProgress());
        return new ProgressInfo()
            .setProgress(updateStatus.getProgress())
            .setIndeterminate(updateStatus.getStatusCode() == INDETERMINATE)
            .setExtras(String.format("%s %d/2", getString(R.string.step), updateStatus.getStep()))
            .setStatus(status);
    }

    private String getString(int id) {
        return context.getString(id);
    }
}

File: app/src/main/java/com/krypton/updater/data/update/UpdateRepository.kt
/*
 * Copyright (C) 2021-2022 AOSP-Krypton Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.krypton.updater.data.update

import android.content.Context
import android.net.Uri

import com.krypton.updater.data.FileCopyStatus
import com.krypton.updater.data.download.DownloadManager
import com.krypton.updater.data.download.DownloadState
import com.krypton.updater.data.savedStateDataStore

import dagger.hilt.android.qualifiers.ApplicationContext

import javax.inject.Inject
import javax.inject.Singleton

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@Singleton
class UpdateRepository @Inject constructor(
    applicationScope: CoroutineScope,
    @ApplicationContext private val context: Context,
    private val updateManager: UpdateManager,
    private val otaFileManager: OTAFileManager,
    private val downloadManager: DownloadManager,
) {

    private val savedStateDataStore = context.savedStateDataStore

    val updateState: StateFlow<UpdateState>
        get() = updateManager.updateState

    val isUpdating: Boolean
        get() = updateManager.isUpdating

    val isUpdatePaused: Boolean
        get() = updateManager.isUpdatePaused

    private val _readyForUpdate = MutableStateFlow(false)
    val readyForUpdate: StateFlow<Boolean>
        get() = _readyForUpdate

    val fileCopyStatus = Channel<FileCopyStatus>(Channel.CONFLATED)

    val supportsUpdateSuspension: Boolean
        get() = updateManager.supportsUpdateSuspension

    init {
        applicationScope.launch {
            downloadManager.downloadState.collect {
                if (it is DownloadState.Finished) {
                    downloadManager.downloadFile?.let { file ->
                        withContext(Dispatchers.IO) {
                            copyOTAFile(Uri.fromFile(file))
                        }
                    }
                } else if (it is DownloadState.Idle && updateState.value is UpdateState.Idle) {
                    _readyForUpdate.value = false
                }
            }
        }
        applicationScope.launch {
            updateState.collect {
                if (it is UpdateState.Finished) saveUpdateFinishedState()
            }
        }
    }

    suspend fun start() {
        withContext(Dispatchers.IO) {
            updateManager.start()
        }
    }

    suspend fun pause() {
        withContext(Dispatchers.IO) {
            updateManager.pause()
        }
    }

    suspend fun resume() {
        withContext(Dispatchers.IO) {
            updateManager.resume()
        }
    }

    suspend fun cancel() {
        withContext(Dispatchers.IO) {
            updateManager.cancel()
        }
    }

    suspend fun reboot() {
        withContext(Dispatchers.IO) {
            updateManager.reboot()
        }
    }

    /**
     * Prepare for local upgrade.
     *
     * @param uri the [Uri] of the update zip file.
     */
    suspend fun copyOTAFile(uri: Uri) {
        updateManager.reset()
        clearSavedUpdateState()
        _readyForUpdate.value = false
        fileCopyStatus.send(FileCopyStatus.Copying)
        val result = withContext(Dispatchers.IO) {
            otaFileManager.copyToOTAPackageDir(uri)
        }
        if (result.isSuccess) {
            fileCopyStatus.send(FileCopyStatus.Success)
        } else {
            fileCopyStatus.send(FileCopyStatus.Failure(result.exceptionOrNull()?.localizedMessage))
        }
        _readyForUpdate.value = result.isSuccess
    }

    fun resetState() {
        _readyForUpdate.value = false
        updateManager.reset()
    }

    private suspend fun saveUpdateFinishedState() {
        savedStateDataStore.updateData {
            it.toBuilder()
                .setUpdateFinished(true)
                .build()
        }
    }

    private suspend fun clearSavedUpdateState() {
        savedStateDataStore.updateData {
            it.toBuilder()
                .clearUpdateFinished()
                .build()
        }
    }
}